server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileManager.java

```diff
@@ -571,4 +583,98 @@ public MetadataUpdateCount getUpdateCount() {
     return metadataUpdateCount.get();
   }
 
+  /**
+   * This is called when metadata file list is determined to be out of sync withe in-memory file
+   * list. Firstly the operation log is dumped. Then metadata is compared again with memory. If
+   * still out of sync, then recovery is performed dependent on the action: 1) log: nothing else is
+   * done 2) logsync: memory is synched with the metadata IFF the operation log agrees with one or
+   * the other 3) metasync: metadata files are reloaded into memory IFF the operation log agrees
+   * with metadata 4) memsync: metadata is reset with the memory files IFF the operation log agrees
+   * with memory
+   *
+   * @param context The server context
+   */
+  public void handleMetadataDiff(ServerContext context) {
+    LogSync action =
+        LogSync.valueOf(tablet.getTableConfiguration().get(Property.TABLE_OPERATION_LOG_RECOVERY));
+    synchronized (tablet) {
+      // always log the operation log regardless of the requested action
+      log.error("Operation log: " + tabletLog.dumpLog());
+
+      // rescan for the tablet metadata
+      var tabletMeta =
+          context.getAmple().readTablet(tablet.getExtent(), TabletMetadata.ColumnType.FILES);
+      if (tabletMeta == null) {
+        String msg = "Tablet " + tablet.getExtent() + " not found in metadata";
+        log.error(msg);
+      } else {
+        Set<StoredTabletFile> metadata = new HashSet<>(tabletMeta.getFiles());
+        Set<StoredTabletFile> memory = new HashSet<>(datafileSizes.keySet());
+        Set<StoredTabletFile> expected = tabletLog.getExpectedFiles();
+
+        // verify we are still out of sync
+        if (metadata.equals(memory)) {
+          log.debug("Metadata and in-memory file list are back in-sync: " + metadata);
+          if (!expected.equals(memory)) {
+            log.error(
+                "Resetting operation log " + expected + " with metadata and memory " + memory);
+            tabletLog.reset(memory);
+          }
+        } else if (!action.equals(LogSync.log)) {
+          if (action.equals(LogSync.logsync)) {
```

Comment:
If the user sets the property to `logsync`, then it gets overwritten below? I'm not sure I understand how/where `log` and `logsync` are used.

Reply:
See comments in the Property or at the top of this method.  "log" means that we simply log the operation log.  logsync means that we will either metasync or memsync depending on which the log agrees with.

Reply:
I think `log` could be removed. It should always log when things don't match, right? I don't see where `log` is ever used. Does `logsync` mean log and then sync? I think `logSync` could be removed also. The property could just be a boolean I think, whether or not to sync the metadata to the tablet memory.

Reply:
Adding in line comments to help explain.

Reply:
I was thinking something similar. I don't know if I'd feel confident picking memsync or metasync. Wondering if we just make the actions either: `NOTHING, FIX`

Reply:
Based on this code, the only thing that matters is whether or not the user picked `logsync` for the property value. If the user uses `memsync` or `metasync`, then nothing will happen.